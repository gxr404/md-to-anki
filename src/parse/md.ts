import { Marked } from 'marked'
import { markedHighlight } from 'marked-highlight'
import { codeToHtml } from 'shiki'

let marked: Marked

async function initMarked() {
  if (marked)
    return marked
  // const highlighter = await initHighlighter()
  marked = new Marked({
    gfm: true,
    breaks: true,
    pedantic: false,
    async: true,
    ...markedHighlight({
      langPrefix: 'core lang-',
      async: true,
      async highlight(code, lang) {
        const langAlias: any = {
          sol: 'solidity',
        }
        const htmlStr = await codeToHtml(code, {
          lang: langAlias[lang] || lang,
          theme: 'vitesse-dark',
        })
        return htmlStr
      },
    }),
  })
  return marked
}

export function getMarked() {
  return marked
}

// let highlighter: HighlighterCore

// async function initHighlighter() {
//   if (highlighter)
//     return highlighter
//   // highlighter = await getHighlighterCore({
//   //   themes: [
//   //     import('shiki/themes/vitesse-dark.mjs'),
//   //   ],
//   //   langs: [
//   //     import('shiki/langs/javascript.mjs'),
//   //     import('shiki/langs/css.mjs'),
//   //     import('shiki/langs/bash.mjs'),
//   //     import('shiki/langs/bat.mjs'),
//   //     import('shiki/langs/c.mjs'),
//   //     import('shiki/langs/coffeescript.mjs'),
//   //     import('shiki/langs/go.mjs'),
//   //     import('shiki/langs/json.mjs'),
//   //     import('shiki/langs/graphql.mjs'),
//   //     import('shiki/langs/java.mjs'),
//   //     import('shiki/langs/less.mjs'),
//   //     import('shiki/langs/markdown.mjs'),
//   //     import('shiki/langs/php.mjs'),
//   //     import('shiki/langs/python.mjs'),
//   //     import('shiki/langs/ruby.mjs'),
//   //     import('shiki/langs/rust.mjs'),
//   //     import('shiki/langs/sass.mjs'),
//   //     import('shiki/langs/stylus.mjs'),
//   //     import('shiki/langs/typescript.mjs'),
//   //     import('shiki/langs/vim.mjs'),
//   //     import('shiki/langs/jsx.mjs'),
//   //     import('shiki/langs/vue.mjs'),
//   //     import('shiki/langs/sql.mjs'),
//   //     import('shiki/langs/csharp.mjs'),
//   //     import('shiki/langs/yaml.mjs'),
//   //   ],
//   //   langAlias: {
//   //     'c#': 'csharp',
//   //     'md': 'markdown',
//   //     'rb': 'ruby',
//   //     'py': 'python',
//   //     'sh': 'bash',
//   //   },
//   //   loadWasm: getWasm,
//   // })
//   const { getHighlighter, bundledLanguages } = await import('shiki')
//   highlighter = await getHighlighter({
//     themes: ['vitesse-dark'],
//     langs: Object.keys(bundledLanguages),
//     // [
//     //   'javascript',
//     //   'css',
//     //   'bash',
//     //   'bat',
//     //   'c',
//     //   'coffeescript',
//     //   'go',
//     //   'json',
//     //   'graphql',
//     //   'java',
//     //   'less',
//     //   'markdown',
//     //   'php',
//     //   'python',
//     //   'ruby',
//     //   'rust',
//     //   'sass',
//     //   'stylus',
//     //   'typescript',
//     //   'vim',
//     //   'jsx',
//     //   'vue',
//     //   'sql',
//     //   'csharp',
//     //   'yaml',
//     //   'http',
//     //   'proto',
//     //   'cpp',
//     //   'solidity',
//     // ],
//     langAlias: {
//       'c#': 'csharp',
//       'md': 'markdown',
//       'rb': 'ruby',
//       'py': 'python',
//       'sh': 'bash',
//       'sol': 'solidity',
//       'c++': 'cpp',
//     },
//   })
//   return highlighter
// }

export async function md2Html(mdStr: string) {
  const marked = await initMarked()
  const htmlStr = marked.parse(mdStr)
  return htmlStr
}
