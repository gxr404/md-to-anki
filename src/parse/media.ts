import path from 'node:path'
import { readFile } from 'node:fs/promises'
import crypto from 'node:crypto'
import axios from 'axios'
import mime from 'mime-types'

import { getExtensionFromUrl } from '../utils'
import type { ICard } from './index'

export const srcReg = /src="([^"]*?)"/g

export interface IMedia {
  data: any
  fileName: string
  srcStr: string
  checksum: string
}

export const mediaList: IMedia[] = []

function addMedia(media: IMedia) {
  const hasMedia = mediaList.some((item: IMedia) => item.checksum === media.checksum)
  if (hasMedia)
    return

  mediaList.push(media)
}
/** 从cards 获取 media信息 */
export async function addMediaToCards(cards: ICard[], sourceFile: string) {
  for (const card of cards) {
    const frontMedia = await mediaParse(card.front, sourceFile)
    const backMedia = await mediaParse(card.back, sourceFile)
    card.front = frontMedia.mdStr
    card.back = backMedia.mdStr
    frontMedia.mediaList.forEach(addMedia)
    backMedia.mediaList.forEach(addMedia)
  }
  return mediaList
}

/** 重置全局的mediaList */
export function resetMediaList() {
  mediaList.splice(0, mediaList.length)
}

export async function mediaParse(mdStr: string, sourceFile: string) {
  const srcArr = mdStr.match(srcReg) || []
  const urlArr = srcArr.map(str => str.replace(srcReg, '$1'))

  const promiseList = urlArr.map(async (url: string) => getMedia(url, sourceFile))
  const mediaList = await Promise.all(promiseList)

  const newMdStr = mdStr.replace(srcReg, (src: string, url: string) => {
    const index = urlArr.indexOf(url)
    return mediaList[index].srcStr
  })

  return {
    mediaList,
    mdStr: newMdStr,
  }
  // str.replace(srcReg, (match, ...args) => {
  //   const promise = asyncFn(match, ...args)
  //   tasks.push(promise)
  // })
}

async function getMedia(urlStr: string, sourceFile: string): Promise<IMedia> {
  let bufferData
  let fileExt
  if (urlStr.startsWith('http')) {
    const resp = await axios.get(urlStr, {
      responseType: 'arraybuffer',
    })
    bufferData = resp.data
    fileExt = getExtensionFromUrl(urlStr)
    if (!fileExt && resp.headers['Content-Type']) {
      const resExt = mime.extension(String(resp.headers['Content-Type']))
      fileExt = resExt || ''
    }
  }
  else {
    const filePath = path.resolve(path.dirname(sourceFile), urlStr)
    fileExt = path.extname(filePath)
    bufferData = await readFile(filePath)
  }

  const checksum = crypto
    .createHash('md5')
    .update(bufferData, 'utf8')
    .digest('hex')
  const fileName = `${checksum}${fileExt}`

  return {
    data: bufferData,
    fileName,
    srcStr: `src="${fileName}"`,
    checksum,
  }
}
