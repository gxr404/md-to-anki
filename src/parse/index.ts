import { readFile } from 'node:fs/promises'
import { getConfig } from '../config'
import { removeMdComments, trimArray } from '../utils'
import { md2Html } from './md'
import { addMediaToCards, mediaList, resetMediaList } from './media'

export async function parseMarkdown(file: string) {
  const mdBuffer = await readFile(file)
  const mdString = removeMdComments(mdBuffer.toString())
  const { rawCards, deckName } = await splitByCards(mdString)
  // console.time('parseCardList')
  const { cards, media } = await parseCardList(rawCards, file)
  // console.timeEnd('parseCardList')

  return {
    cards,
    media,
    deckName,
  }
}

export interface ICard {
  front: string
  back: string
  tags: string[]
}

export async function splitByCards(mdString: string) {
  const config = getConfig()
  const rawCards = mdString
    .split(new RegExp(config.card.separator, 'm'))
    .map((line: string) => line.trim())
  return {
    deckName: getDeckName(rawCards),
    rawCards,
  }
}

function getDeckName(rawCards: string[]) {
  const config = getConfig()
  const titleReg = new RegExp(config.deck.titleSeparator)
  let deckName = rawCards
    .find((str: string) => titleReg.test(str)) ?? ''
  deckName = deckName.split('\n')?.[0] || ''
  if (!deckName)
    return config.deck.defaultName

  return deckName.replace(/(#\s|\n)/g, '').trim()
}

export async function parseCardList(rawCards: string[], sourceFile: string) {
  resetMediaList()
  const dirtyCards = await Promise.all(
    rawCards.map(async (str: string) => {
      // console.time(`parseCard#${i}`)
      const card = await parseCard(str)
      // console.timeEnd(`parseCard#${i}`)

      if (card)
        await addMediaToCards([card], sourceFile)
      return card
    }),
  )

  const cards = dirtyCards
    .filter(card => Boolean(card))
    // card should have front and back sides
    .filter(card => card?.front && card?.back) as ICard[]
  return {
    cards,
    media: mediaList,
  }
}

export async function parseCard(cardStr: string = '') {
  const config = getConfig()
  const splitCardReg = new RegExp(`^${config.card.frontBackSeparator}$`, 'm')
  // 卡片的内容 按每一行分割
  const cardLines = cardStr
    .split(splitCardReg)
    .map((item: string) => item.split('\n'))
    .map((arr: string[]) => arr.map((str: string) => str.trimEnd()))

  // 不允许卡片只有正面(not allowed cards with only front side)
  if (cardLines.length === 1 && !cardLines[0].filter(Boolean).length)
    return null

  const { front, back, tags } = parseCardLines(cardLines)

  // 一般都是转html不存在不转的情况
  // if (!this.options.convertToHtml) {
  //   return new Card(front.join(), back.join(), tags);
  // }
  const frontHtml = await md2Html(front.join('\n'))
  const backHtml = await md2Html(back.join('\n'))

  return {
    front: frontHtml,
    back: backHtml,
    tags,
  }
}

function parseTags(line: string) {
  const config = getConfig()
  const tagReg = new RegExp(config.card.tagPattern)
  const data = line.split(' ')
    .map((str: string) => str.trim())
    .map((str) => {
      const parts = tagReg.exec(str)
      return parts?.[1] ?? ''
    })
    .filter(Boolean)

  return data
}

function parseCardLines(cardLines: string[][]) {
  const config = getConfig()
  const tagReg = new RegExp(config.card.tagPattern)

  const front: string[] = []
  const back: string[] = []
  const tags: string[] = []

  function fillTagAndBack(line: string) {
    if (tagReg.test(line) && line) {
      tags.push(...parseTags(line))
      return
    }
    // back 第一个元素如果是空 则跳过
    if (back.length === 0 && !line)
      return
    back.push(line)
  }

  // 没设置卡片分割符 默认 正面为第一行，其余为反面
  if (cardLines.length === 1) {
    trimArray(cardLines[0])
      .forEach((line, index) => {
        // 默认 正面为第一行
        if (index === 0) {
          front.push(line)
          return
        }
        fillTagAndBack(line)
      })
  }
  else {
    front.push(...cardLines[0])
    trimArray(cardLines[1])
      .forEach((line: string) => {
        fillTagAndBack(line)
      })
  }

  return {
    front: trimArray(front),
    back: trimArray(back),
    tags: trimArray(tags),
  }
}

export * from './media'
