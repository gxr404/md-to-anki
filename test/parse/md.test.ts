import { afterEach, describe, expect, it, vi } from 'vitest'
import { getMarked, md2Html } from '../../src/parse/md'

afterEach(() => {
  vi.restoreAllMocks()
})

describe('parse markdown', () => {
  it('parse code', async () => {
    const htmlStr = await md2Html('\n```js\nvar a = 123\n```\n')
    expect(htmlStr).toMatchSnapshot()
  })
  it('parse alias sol code', async () => {
    const htmlStr = await md2Html('\n```sol\npragma solidity ^0.8.4;\n```\n')
    expect(htmlStr).toMatchSnapshot()
  })
  it('parse error code', async () => {
    const markedParse = () => {
      throw new Error('cannot parse')
    }
    const marked = getMarked()
    vi.spyOn(marked, 'parse').mockImplementation(markedParse)
    let callErr: any
    try {
      await md2Html('\n```js\nvar a = 123\n```\n')
    }
    catch (err: any) {
      callErr = err
    }
    expect(callErr?.message).toBe('cannot parse')
  })
})
