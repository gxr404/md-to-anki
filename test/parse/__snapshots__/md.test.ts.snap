// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse markdown > parse alias sol code 1`] = `
"<pre><code class="core lang-sol"><pre class="shiki vitesse-dark" style="background-color:#121212;color:#dbd7caee" tabindex="0"><code><span class="line"><span style="color:#4D9375">pragma</span><span style="color:#4D9375"> solidity</span><span style="color:#C99076"> ^0.8.4</span><span style="color:#666666">;</span></span></code></pre>
</code></pre>"
`;

exports[`parse markdown > parse code 1`] = `
"<pre><code class="core lang-js"><pre class="shiki vitesse-dark" style="background-color:#121212;color:#dbd7caee" tabindex="0"><code><span class="line"><span style="color:#CB7676">var</span><span style="color:#BD976A"> a</span><span style="color:#666666"> =</span><span style="color:#4C9A91"> 123</span></span></code></pre>
</code></pre>"
`;
